app/models/smart_proxy.rb

```diff
@@ -82,18 +82,34 @@ def refresh
   end
 
   def taxonomy_foreign_conditions
-    feature_names = self.features.pluck(:name)
-    conditions                      = {}
-    if feature_names.include?('Puppet') && feature_names.include?('Puppet CA')
+    conditions = {}
+    if has_puppet? && has_puppet_ca?
       conditions = "puppet_proxy_id = #{id} OR puppet_ca_proxy_id = #{id}"
-    elsif feature_names.include?('Puppet')
+    elsif has_puppet?
       conditions[:puppet_proxy_id] = id
-    elsif feature_names.include?('Puppet CA')
+    elsif has_puppet_ca?
       conditions[:puppet_ca_proxy_id] = id
     end
     conditions
   end
 
+  # Helper methods for features - e.g. has_puppet? has_dns?
+  def method_missing(method, *args, &block)
```

Comment:
I see the comment on the redmine ticket http://projects.theforeman.org/issues/10287#note-1 - 
but this is method_missing, the same kind of problems won't happen here.  It's not done at class evaluation like that dynamic scope code was.

And really, those dynamic scopes could have just been converted to method_missing instead of deprecated and it would've sovled the problems.

I *really* like this way...